libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2023 Matthieu Bouron <matthieu.bouron@gmail.com>
+ * Copyright 2023 Clément Bœsch <u pkh.me>
+ * Copyright 2023 Nope Foundry
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+float u32tof32(highp uint x)
+{
+    uint y = (0x7FU << 23U) | (x >> 9U);
+    return uintBitsToFloat(y) - 1.0;
+}
+
+highp uint lowbias32(highp uint x)
+{
+    x ^= x >> 16;
+    x *= 0x7feb352dU;
+    x ^= x >> 15;
+    x *= 0x846ca68bU;
+    x ^= x >> 16;
+    return x;
+}
+
+uint hash(uvec2 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y));
+}
+
+uint hash(uvec3 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y ^ lowbias32(x.z)));
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+/*
+ * References:
+ * - https://web.archive.org/web/20200618040237/https://mrl.nyu.edu/~perlin/paper445.pdf
+ * - https://web.archive.org/web/20180616011332/http://www.scratchapixel.com/lessons/procedural-generation-virtual-worlds%20/perlin-noise-part-2
+ */
+vec3 random_grad(uvec3 x)
+{
+    uint h = hash(x);
+    float r0 = u32tof32(0x00010001U * (h      & 0xffffU));
+    float r1 = u32tof32(0x00010001U * (h>>16U & 0xffffU));
+    float theta = acos(2.0 * r0 - 1.0);
+    float phi = ngli_tau * r1;
+    return vec3(cos(phi) * sin(theta), sin(phi) * sin(theta), cos(theta));
+}
```

Comment:
Might want to add a few comments here...

```glsl
/*
 * Generate a random point on a sphere (with uniform distribution)
 * 
 * References:
 * - https://web.archive.org/web/20200618040237/https://mrl.nyu.edu/~perlin/paper445.pdf
 * - https://web.archive.org/web/20180616011332/http://www.scratchapixel.com/lessons/procedural-generation-virtual-worlds%20/perlin-noise-part-2
 */
vec3 random_grad(uvec3 x)
{
    /* Generate 2 random floats in [0;1] using 16-bit of information for each */
    uint h = hash(x);
    float r0 = u32tof32(0x00010001U * (h      & 0xffffU));
    float r1 = u32tof32(0x00010001U * (h>>16U & 0xffffU));

    float theta = acos(2.0 * r0 - 1.0); // use the first random for the polar angle (latitude)
    float phi = ngli_tau * r1; // use the 2nd random for the azimuth (longitude)
    return vec3(cos(phi) * sin(theta), sin(phi) * sin(theta), cos(theta));
}
```

Reply:
Done.